src/wallet/wallet.cpp

```diff
@@ -4135,3 +4138,106 @@ bool CWalletTx::AcceptToMemoryPool(const CAmount& nAbsurdFee, CValidationState&
+    }
+}
+
+std::vector<CTxDestination> GetAllDestinationsForKey(const CPubKey& key)
+{
+    CKeyID keyid = key.GetID();
+    if (key.IsCompressed()) {
+        CTxDestination segwit = WitnessV0KeyHash(keyid);
+        CTxDestination p2sh = CScriptID(GetScriptForDestination(segwit));
+        return std::vector<CTxDestination>{std::move(keyid), std::move(p2sh), std::move(segwit)};
+    } else {
+        return std::vector<CTxDestination>{std::move(keyid)};
+    }
+}
+
+CTxDestination CWallet::AddDestinationForKey(const CPubKey& key, OutputType type)
+{
+    switch (type) {
+    case OUTPUT_TYPE_LEGACY: return key.GetID();
+    case OUTPUT_TYPE_P2SH:
+    case OUTPUT_TYPE_BECH32: {
+        CTxDestination witdest = WitnessV0KeyHash(key.GetID());
+        CScript witprog = GetScriptForDestination(witdest);
+        // Check if the resulting program is solvable (i.e. doesn't use an uncompressed key)
+        if (!IsSolvable(*this, witprog)) return key.GetID();
+        // Even though CBasicKeyStore knows about P2WPKH redeemscripts implicitly, add it explicitly
+        // so that downgrades to software without this are supported (as long as the wallet file is
+        // up to date).
+        AddCScript(witprog);
+        if (type == OUTPUT_TYPE_P2SH) {
+            return CScriptID(witprog);
+        } else {
+            return witdest;
+        }
+    }
+    default: assert(false);
+    }
+}
+
+CTxDestination CWallet::AddDestinationForScript(const CScript& script, OutputType type)
+{
+    // Note that scripts over 520 bytes are not yet supported.
+    switch (type) {
+    case OUTPUT_TYPE_LEGACY:
+    case OUTPUT_TYPE_P2SH:
+        return CScriptID(script);
+    case OUTPUT_TYPE_BECH32: {
+        WitnessV0ScriptHash hash;
+        CSHA256().Write(script.data(), script.size()).Finalize(hash.begin());
+        CTxDestination witdest = hash;
+        CScript witprog = GetScriptForDestination(witdest);
+        // Check if the resulting program is solvable (i.e. doesn't use an uncompressed key)
+        if (!IsSolvable(*this, witprog)) return CScriptID(script);
+        AddCScript(witprog);
+        // Add the redeemscript, so that P2WSH and P2SH-P2WSH outputs are recognized as ours.
+        return witdest;
+    }
+    default: assert(false);
+    }
+}
+
+void CWallet::AddRelatedScripts(const CScript& scriptPubKey, const CKeyID& keyid)
+{
+    CTxDestination segwit = WitnessV0KeyHash(keyid);
+    CTxDestination p2sh = CScriptID(GetScriptForDestination(segwit));
+    // For SegWit scriptPubKeys, add the redeemscript explicitly to the wallet, so that
+    // older software versions (without implicit scripts) correctly treat these as ours.
+    if (scriptPubKey == GetScriptForDestination(segwit) || scriptPubKey == GetScriptForDestination(p2sh)) {
+        CTxDestination witdest = WitnessV0KeyHash(keyid);
+        CScript witprog = GetScriptForDestination(witdest);
+        if (IsSolvable(*this, witprog)) {
+            AddCScript(witprog);
```

Comment:
In commit "Support downgrading after recovered keypool witness keys"

No test seems to fail if this is commented out. Since this is only needed for downgrades it'd be good to have some kind of check to know if it stops working.

Reply:
Likewise, this is untestable because it only affects downgrading.